### ISSUE TITLE
[Feature Request] Array access shorthand for endpoints of a Domain

### ISSUE THREAD

**Issue by jeremiah-corrado** [chapel-lang/chapel#20409]
## Proposal

Introduce a new symbol into the array access syntax which acts as a shorthand for either the `low` or `high` index along the relevant dimension. 

The new symbol could be used both in direct array access, i.e. retrieving a single value or dimension, and array slicing using ranges, i.e. getting a range of values along a dimension. 

In the following examples and explanation I'll use the carrot (`^`) symbol; however I'd be open to swapping it for something else. (There is a potential for confusion when the endpoints of a range are defined in terms of some exponential expression like: `0..5^-2`, but I don't think there would be a parsing/syntax conflict)

## Syntax examples

The details of this proposal are best explained using examples. The following should be pretty comprehensive:

```chapel
// set all but the first and last elements of an array to zero
a[^+1..^-1] = 0;

// set an inner slice to zero (ranging from the 4th element to the 7th element from the end)
a[^+4..^-7] = 0;

// set the 5th element of an array to zero
a[^+5] = 0;

// set the first and last elements to zero (respectively)
a[^+] = 0;
a[^-] = 0;

// set the second to last column of a 2D array to zero
a[.., ^-2] = 0;

// swap the first and last rows of a 2D array
a[^+, ..] <=> a[^-, ..]
```

In the above examples, the `^-` syntax (to index from the end of an array) is of course more useful, as `^+` is standing in for `0` (assuming the array indexing is zero-based). However, I think it would be beneficial to include both for the sake of completeness, and because `^+` would allow users to write code which is more generic over different starting indices.

The syntax for indexing into the entire array would technically be expanded to include the following; however, I doubt these would be used in practice:
```chapel
a[^+..^-] = 0;
a[^+..] = 0;
a[..^-] = 0;
```

## Motivating Example

Recently, a realistic example of how this could be useful came up while porting some Finite Difference code from Python to Chapel. The following Python function sets the normal "derivative" to zero on the border of a 2D numpy array. (It essentially copies the values which are 1 element away from the edge onto the nearest edge).
```python
def zero_deriv_border(p):
    p[0, :] = p[1, :]   # dp/dx = 0 (left border)
    p[:, 0] = p[:, 1]   # dp/dy = 0 (bottom border)
    p[-1, :] = p[-2, :] # dp/dx = 0 (right border)
    p[:, -1] = p[:, -2] # dp/dy = 0 (top border)
```
Currently, this translates into the following Chapel code (assuming p has zero-based indexing):
```chapel
proc zeroDerivBorder(p) {
    p[0, ..] = p[1, ..]; // dp/dx = 0 (left border)
    p[.., 0] = p[.., 1]; // dp/dy = 0 (bottom border)
    p[p.domain.dim(0).high, ..] = [p.domain.dim(0).high-1, ..]; // dp/dx = 0 (right border)
    p[.., p.domain.dim(1).high] = [.., p.domain.dim(1).high-1]; // dp/dy = 0 (top border)
}
```
The second two lines are unfortunately long, and indicate to me that it would be nice to have something similar to the python syntax available in chapel arrays. 

Here is the same procedure with the proposed change (which also works with non-zero-based indexing):
```chapel
proc zeroDerivBorder(p) {
    p[^+, ..] = p[^+1, ..]; // dp/dx = 0 (left border)
    p[.., ^+] = p[.., ^+1]; // dp/dy = 0 (bottom border)
    p[^-, ..] = p[^-1, ..]; // dp/dx = 0 (right border)
    p[.., ^-] = p[.., ^-1]; // dp/dy = 0 (top border)
}
```

### Invoking `^+` or `^-` Outside of an Indexing Context:

This proposal would require that ranges support the new syntax; however this raises the question of what the new syntax would mean outside the context of array indexing. 

For example what would the following code print?
```chapel
for i in 0..^- {
    writeln(i);
}
```

I think the simplest solution would be to throw an error whenever the new syntax is invoked outside the context of array indexing. I would also be open to other suggestions.

**Comment by jeremiah-corrado** [comment]
Pinging @bradcray and @vasslitvinov due to your recent work/interest in ranges.

**Comment by bradcray** [comment]
I like the idea of this feature, though believe there will be challenges to implementing it since:
* most range bounds expressions are legal expressions outside of the range context
* most range expressions are legal range expressions outside of the slicing context
That said, if we had this feature, I would use it (and have found myself regretting that we don't have it at times.

When we were kicking this around interactively the other day, I suggested that `_` could be an alternate placeholder for the bound because (a) it's already in the language in certain places, (b) it looks visually like a placeholder ("fill in the blank"), (c) it's not a legal identifier, but is identifier-like.  That said, this use of it would differ somewhat from other uses (such as "drop this tuple component on the floor"—though I can also argue a certain sense of compatibility).

Another mildly related concept that we were kicking around was the use of open unbounded ranges to do this for the common case of +1/-1.  E.g., imagine:

```chapel
var A: [1..n] real;
A[..<]  // A[1..n-1]
A[>..] // A[2..n]
A[>..<] // A2..-n1]
```

Beyond the obvious "only works for a single index", this has similar downsides as the proposal above that it makes certain range expressions that would typically be nonsensical suddenly have meaning when used in a slicing context.  But they could also be syntactically constrained to only occur within slicing expressions.

I don't think this solution is complete due to the "single index" nature, but if we had both features, I'd probably prefer to use this one if my offsets were just +1 / -1.

**Comment by vasslitvinov** [comment]
This can be made to work. Checking whether these "bounds" are used correctly can be done during resolution, not necessarily during parsing.

In terms of syntax, we could combine the `><` and the `^` ideas into "seriously-open" ranges:

```chpl
A[..5<]  // A[1..n-5]
A[>3..]  // A[4..n]
A[>-3..] // A[n-3..n]
```

The shortcoming is that open ranges are still ranges. Whereas Jeremiah's carrot syntax enables rank-changing slices.

The `#` operator is somewhat related, and we could beef it up to allow tuple arguments (each for a separate dimension), if it is not there yet. However, it counts only from one end of the range and perhaps cannot (be made to) create slices.

**Comment by jeremiah-corrado** [comment]
> I don't think this solution is complete due to the "single index" nature, but if we had both features, I'd probably prefer to use this one if my offsets were just +1 / -1.

I agree, this syntax looks much nicer than the `^+` and `^-` for single index offsets. I'd be in favor of supporting both proposals.

> In terms of syntax, we could combine the >< and the ^ ideas into "seriously-open" ranges:

I did consider this; however, I am somewhat hesitant because I wanted to have a symbol which very clearly acts as a direct stand-in for the first/last indices.

Let me know if I am misunderstanding, but it looks like `A[>..]` and `A[>1..]` would both mean `A[2..]`. However I think some users might interpret `>` as always acting as a shorthand for the second index of the array, so they might expect `A[>1..]` to mean `A[3..]` (i.e. the second index, plus 1). 

So I think it would be beneficial to support both features separately. 

> When we were kicking this around interactively the other day, I suggested that _ could be an alternate placeholder for the bound

I like `_` as well. I think it looks pretty nice in the "motivating example" from above:
```chapel
a[_+, ..] = a[_+1, ..];
a[_-, ..] = a[_-1, ..];
a[.., _+] = a[.., _+1];
a[.., _-] = a[.., _-1];
```
___
Another possibility would be supporting two separate symbols: maybe `_` for the lower bound and `^` for the upper bound. This way we could support "seriously-open" ranges as Vass suggested. I.e. the upper bound could be queried before the `..` and the lower bound after the `..`, allowing users to easily slice out boundary regions of an array:
```chapel
A[^-3..] // A[n-3..n]
A[.._+5] // A[1..6]
```
